system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -0,0 +1,1149 @@
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include <chrono>
+#include <curl/curl.h>
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "basesecurity.hpp"
+#include "authmap.ipp"
+#include "daldap.hpp"
+#include "dasess.hpp"
+#include "daliKVStore.hpp"
+#include "jsecrets.hpp"
+
+// Need to temporarily disable the verify macro because the jwt-cpp code has
+// a verify() function that conflicts
+#ifdef verify
+    #define JWT_HAS_VERIFY_MACRO
+    #define OLD_VERIFY verify
+    #undef verify
+#endif
+
+#include "jwt-cpp/jwt.h"
+
+#include "nlohmann/json.hpp"
+
+#include "jwtCache.hpp"
+#include "jwtEndpoint.hpp"
+#include "jwtSecurity.hpp"
+
+static const char* DALI_KVSTORE_STORE_NAME = "JWTAuth";
+static const char* DALI_KVSTORE_NAMESPACE = "Tokens";
+static const char* DALI_KVSTORE_KEY = "Token";
+
+// The number of milliseconds between checking the key file for updates; this
+// means changes to the file won't be noticed until at most
+// KEY_FILE_CHECK_INTERVAL_MS milliseconds elapse
+#define KEY_FILE_CHECK_INTERVAL_MS 10000
+
+static JWTUserCache     gUserCache;
+
+/**
+ * Class used to temporarily hold scope permissions parsed from tokens.
+ */
+class ScopePermission
+{
+    public:
+
+        ScopePermission(const std::string& _jwtClaim, const std::string& _scopePattern)
+        :   scopePattern(_scopePattern), claim(_jwtClaim)
+        {
+        }
+
+        bool operator< (const ScopePermission& other) const
+        {
+            return scopePattern < other.scopePattern;
+        }
+
+        std::string     scopePattern;
+        std::string     claim;
+};
+
+typedef std::vector<ScopePermission> CollectedScopePerms;
+
+class CJwtSecurityManager : implements IDaliLdapConnection, public CBaseSecurityManager
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CJwtSecurityManager(const char* serviceName, IPropertyTree* secMgrCfg, IPropertyTree* bindConfig)
+    : CBaseSecurityManager(serviceName, (IPropertyTree*)nullptr), keyIsPublicKey(false)
+    {
+        defaultPermAccess = SecAccess_Full;
+        defaultWUAccess = SecAccess_Full;
+        defaultFileAccess = SecAccess_Full;
+
+        if (secMgrCfg != nullptr)
+        {
+            clientID = secMgrCfg->queryProp("@clientID");
+            loginEndpoint = secMgrCfg->queryProp("@loginEndpoint");
+            refreshEndpoint = secMgrCfg->queryProp("@refreshEndpoint");
+            secretsPath = secMgrCfg->queryProp("@secretsPath");
```

Comment:
Minor: This isn't really a path, more a name.

Reply:
Renamed throughout, including in the bare-metal configmgr option.